pkg/kubelet/nodelease/controller.go

```diff
@@ -0,0 +1,179 @@
+	holderIdentity             string
+	leaseDurationSeconds       int32
+	renewInterval              time.Duration
+	clock                      clock.Clock
+	onRepeatedHeartbeatFailure func()
+}
+
+// NewController constructs and returns a controller
+func NewController(clock clock.Clock, client clientset.Interface, holderIdentity string, leaseDurationSeconds int32, onRepeatedHeartbeatFailure func()) Controller {
+	var leaseClient coordclientset.LeaseInterface
+	if client != nil {
+		leaseClient = client.CoordinationV1beta1().Leases(corev1.NamespaceNodeLease)
+	}
+	return &controller{
+		client:               leaseClient,
+		holderIdentity:       holderIdentity,
+		leaseDurationSeconds: leaseDurationSeconds,
+		renewInterval:        time.Duration(float64(time.Duration(leaseDurationSeconds)*time.Second) * renewFactor),
+		clock:                clock,
+		onRepeatedHeartbeatFailure: onRepeatedHeartbeatFailure,
+	}
+}
+
+// Run runs the controller
+func (c *controller) Run(stopCh <-chan struct{}) {
+	if c.client == nil {
+		glog.Infof("node lease controller has nil client, will not claim or renew leases")
+		return
+	}
+	wait.Until(c.sync, c.renewInterval, stopCh)
+}
+
+func (c *controller) sync() {
+	lease, created := c.backoffEnsureLease()
+	// we don't need to update the lease if we just created it
+	if !created {
+		c.retryUpdateLease(lease)
+	}
+}
+
+// backoffEnsureLease attempts to create the lease if it does not exist,
+// and uses exponentially increasing waits to prevent overloading the API server
+// with retries. Returns the lease, and true if this call created the lease,
+// false otherwise.
+func (c *controller) backoffEnsureLease() (*coordv1beta1.Lease, bool) {
+	var (
+		lease   *coordv1beta1.Lease
+		created bool
+		err     error
+	)
+	sleep := 100 * time.Millisecond
+	for {
+		lease, created, err = c.ensureLease()
+		if err == nil {
+			break
+		}
+		sleep = minDuration(2*sleep, maxBackoff)
+		// log error
+		glog.Errorf("failed to ensure node lease exists, will retry in %v, error: %v", sleep, err)
+		// backoff wait
+		c.clock.Sleep(sleep)
+	}
+	return lease, created
+}
+
+// ensureLease creates the lease if it does not exist. Returns the lease and
+// a bool (true if this call created the lease), or any error that occurs.
+func (c *controller) ensureLease() (*coordv1beta1.Lease, bool, error) {
+	lease, err := c.client.Get(c.holderIdentity, metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		// lease does not exist, create it
+		lease, err := c.client.Create(c.newLease(nil))
+		if err != nil {
+			return nil, false, err
+		}
+		return lease, true, nil
+	} else if err != nil {
+		// unexpected error getting lease
+		return nil, false, err
+	}
+	// lease already existed
+	return lease, false, nil
+}
+
+// retryUpdateLease attempts to update the lease for maxUpdateRetries,
+// call this once you're sure the lease has been created
+func (c *controller) retryUpdateLease(base *coordv1beta1.Lease) {
```

Comment:
I know this is copied from `retryUpdateNodeStatus`, but just FYI, we've found before that there were multiple layers of retries..and that was probably unnecessary.  https://github.com/kubernetes/node-problem-detector/issues/124#issuecomment-311449714

Reply:
Noted. Is the client advanced enough now that we could remove onRepeatedHeartbeatFailure too?
@liggitt?

Reply:
No. That is required to force close dead TCP connections.

Reply:
Maybe we can lower the `maxUpdateRetries` later in a separate PR.